db/commitlog/commitlog.cc

```diff
@@ -1379,18 +1439,69 @@ future<> db::commitlog::segment_manager::message_loop() {
         // TODO: ordering of which we issue first.
 
         if (!messages.empty()) {
+            tasks = do_writes(std::move(tasks), messages);
             message_partition mp(messages, shutdown_request_index);
             if (!mp.empty()) {
                 shutdown = true;
             }
         }
 
+        tasks = do_periodic_syncs(std::move(tasks));
+
         assert(messages.empty());
     }
 
     co_await std::move(tasks);
 }
 
+future<> db::commitlog::segment_manager::do_writes(future<> next, std::vector<message>& messages) {
+    message_partition mp(messages, write_request_index, [](auto& m1, auto& m2) {
+        auto& f1 = std::get<write_request>(m1.payload);
+        auto& f2 = std::get<write_request>(m2.payload);
+        return std::tie(f1.segment->_desc.id, f1.file_offset) < std::tie(f2.segment->_desc.id, f2.file_offset);
+    });
+
+    for (auto&& m : mp) {
+        auto& p = std::get<write_request>(m.payload);
+
+        next = [](future<> f, write_request p) -> future<> {
+            auto s = p.segment;
+            try {
+                co_await s->perform_write_request(std::move(p));
+            } catch (...) {
+                // Redundant. Maybe make write noexcept? Alternative remove logging
+                // from it.
+                clogger.error("Error processing write message: {}", std::current_exception());
+            }
+            s = {};
+            co_await std::move(f);
+        }(std::move(next), std::move(p));
+    }
+
+    return next;
+}
+
+future<> db::commitlog::segment_manager::do_periodic_syncs(future<> next) {
+    auto def_copy = _segments;
+    for (auto& s : def_copy) {
+        if (s->must_sync()) {
+            next = [](future<> f, sseg_ptr s) -> future<> {
+                try {
+                    auto wr = s->make_write_request(true);
+                    co_await s->perform_write_request(std::move(wr));
+                } catch (...) {
+                    clogger.error("Error during periodic sync: {}", std::current_exception());
+                }
+                s = {};                
+                co_await std::move(f);
+            }(std::move(next), s);
+        }
```

Comment:
I'm totally confused by this immediately-invoked coroutine that is then immediately co_awaited later on.

Is there no simpler way to write it?

Reply:
Actually, this is slightly broken. It is supposed to be a "return next" below (line 1501), not co_await std::move(next).
The idea here is twofold: 
1.) We are using coroutine/future chaining to build the waitable queue of things we are doing in the background. Other options would be maybe a set of futures or similar, downside being secondary storage + need to prune/gc. 
The risk here of course is instead potentially keeping a coroutine frame alive longer than required....
2.) Avoid making the surrounding function have a coroutine frame (thus borkked with the co_await). Borderline premature opt.

Reply:
Now consistent.